### ISSUE TITLE
lifecycle_node causing launch to hang at shutdown

### ISSUE THREAD

**Issue by pbaughman** [ros2/launch_ros#102]
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 18.04 -->
- Installation type:
  - Source
- Version or commit hash:
  - Launch: c644dfa8e1f1433e200b76eb52ba22f392838f48
  - Launch Ros: 84cdecd361a3ea274f77f0c1ec2f2c02097fe372
- DDS implementation:
  - Fast-RTPS

I'm not sure if this issue is launch or launch_ros related.  I've got another issue open here: https://github.com/ros2/launch/issues/359

It appears that a lifecycle_node can deadlock with the launch service when shutting down.  Both the launch service and the lifecycle_node are waiting on the launch service's _process_one_event to return.  I'm looking for tips for how to continue debugging this.

#### Steps to reproduce issue

I can get this to happen by running `ros2 test lifecycle/test/test_lifecycle.py` in a loop.  Eventually the test will hang after the active tests and before the post-shutdown tests run.

I can determine that launch_testing is stuck in self._launch_service.run()

I can determine the launch service is stuck [here](https://github.com/ros2/launch/blob/2f32b9fa94a5cd441271a9b23850db18b9c7d28e/launch/launch/launch_service.py#L369-L379)

Furthermore, I can print some information about what's in the entity_futures list:

```
    <Task pending coro=<LaunchService._process_one_event() running at /home/pete.baughman/gc/apex_ws/build/launch/launch/launch_service.py:286> wait_for=<Future
 pending cb=[Task._wakeup()] created at /usr/lib/python3.6/asyncio/base_events.py:295> created at /home/pete.baughman/gc/apex_ws/build/launch/launch/launch_service.py:355>
      [
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/launch_service.py, line 414 in run>,
        <FrameSummary file /usr/lib/python3.6/asyncio/base_events.py, line 471 in run_until_complete>,
        <FrameSummary file /usr/lib/python3.6/asyncio/base_events.py, line 438 in run_forever>,
        <FrameSummary file /usr/lib/python3.6/asyncio/base_events.py, line 1451 in _run_once>,
        <FrameSummary file /usr/lib/python3.6/asyncio/events.py, line 145 in _run>,
        <FrameSummary file /usr/lib/python3.6/asyncio/tasks.py, line 261 in _wakeup>,
        <FrameSummary file /usr/lib/python3.6/asyncio/tasks.py, line 180 in _step>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/launch_service.py, line 355 in run_async>
      ]
    <Future pending cb=[_chain_future.<locals>._call_check_cancel() at /usr/lib/python3.6/asyncio/futures.py:403] created at /usr/lib/python3.6/asyncio/base_events.py:295>
      [
        <FrameSummary file /usr/lib/python3.6/asyncio/tasks.py, line 180 in _step>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/launch_service.py, line 287 in _process_one_event>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/launch_service.py, line 307 in __process_event>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/utilities/visit_all_entities_and_collect_futures_impl.py, line 38 in visit_all_entities_and_collect_futures>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/action.py, line 89 in visit>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch/launch/actions/opaque_function.py, line 75 in execute>,
        <FrameSummary file /home/pete.baughman/gc/apex_ws/build/launch_ros/launch_ros/actions/lifecycle_node.py, line 101 in _on_change_state_event>,
        <FrameSummary file /usr/lib/python3.6/asyncio/base_events.py, line 655 in run_in_executor>,
        <FrameSummary file /usr/lib/python3.6/asyncio/futures.py, line 431 in wrap_future>,
        <FrameSummary file /usr/lib/python3.6/asyncio/base_events.py, line 295 in create_future>
      ]
```

I'm not so good at asyncio futures, but it looks like I've got one task pending that was [created here](https://github.com/ros2/launch/blob/2f32b9fa94a5cd441271a9b23850db18b9c7d28e/launch/launch/launch_service.py#L359) and another future in the event queue that came from a launch_ros/actions/lifecycle_node that's also waiting for _process_one_event in the launch_service.

Does any of this look familiar to anybody?  Any tips for how to continue debugging this hang?

**Comment by pbaughman** [comment]
I spoke to @wjwwood and we suspect that things are getting hung [here](https://github.com/ros2/launch_ros/blob/master/launch_ros/launch_ros/actions/lifecycle_node.py#L77-L92)

The lifecycle node checks that a service is available and then tries to call it apparently without a timeout.  If the service goes away after the check but before the call, we can maybe hang forever.

I'm trying to reproduce my issue again with additional instrumentation to confirm this hypothesis

**Comment by pbaughman** [comment]
@wjwwood It looks like you were right.  I added some print statements to LifecycleNode:

```
    def _call_change_state(self, request, context: launch.LaunchContext):
        while not self.__rclpy_change_state_client.wait_for_service(timeout_sec=1.0):
            if context.is_shutdown:
                self.___logger.warning(
                    "Abandoning wait for the '{}' service, due to shutdown.".format(
                        self.__rclpy_change_state_client.srv_name),
                )
                return
        import time; time.sleep(0.5)
        print("Calling the service {}".format(request))
        response = self.__rclpy_change_state_client.call(request)
        print("Finished calling service got response {}".format(response))
        if not response.success:
            self.__logger.error(
                "Failed to make transition '{}' for LifecycleNode '{}'".format(
                    ChangeState.valid_transitions[request.transition.id],
                    self.node_name,
                )
            )
```

And finally got this to hang:
```
[lifecycle_listener-2] [INFO] [listener]: notify callback: Transition from state inactive to cleaningup
[lifecycle_listener-2] [INFO] [listener]: notify callback: Transition from state cleaningup to unconfigured
Finished calling service got response lifecycle_msgs.srv.ChangeState_Response(success=True)
Calling the service lifecycle_msgs.srv.ChangeState_Request(transition=lifecycle_msgs.msg.Transition(id=5, label=''))
[lifecycle_talker-1] [INFO] [lc_talker]: on shutdown is called from state unconfigured.
[lifecycle_listener-2] [INFO] [listener]: notify callback: Transition from state unconfigured to shuttingdown
[lifecycle_listener-2] [INFO] [listener]: notify callback: Transition from state shuttingdown to finalized
ok

----------------------------------------------------------------------
Ran 1 test in 4.858s

OK
[INFO] [lifecycle_listener-2]: sending signal 'SIGINT' to process[lifecycle_listener-2]
[INFO] [lifecycle_talker-1]: sending signal 'SIGINT' to process[lifecycle_talker-1]
[lifecycle_listener-2] [INFO] [rclcpp]: signal_handler(signal_value=2)
[lifecycle_talker-1] [INFO] [rclcpp]: signal_handler(signal_value=2)
[INFO] [lifecycle_listener-2]: process has finished cleanly [pid 13315]
[INFO] [lifecycle_talker-1]: process has finished cleanly [pid 13314]
```

As you can see, the final "Calling the service" does not have a matching "Finished calling service" line.
